src/record_ts.h

```diff
@@ -0,0 +1,531 @@
+
+    // Enqueue a function to be run on the host once backend computation is done
+    void enqueue_host_func(void (*callback)(void *), void *payload) override;
+
+    /// LLVM: reduce a variable that was previously expanded due to
+    /// dr.ReduceOp.Expand
+    void reduce_expanded(VarType vt, ReduceOp reduce_op, void *data,
+                         uint32_t exp, uint32_t size) override;
+
+    /**
+     * This function is called every time a pointer is freed using \ref
+     * jitc_free. It records the operation and removes the mapping from that
+     * pointer to the recorded variable.
+     * If the pointer is reused later by another call to \ref jitc_malloc, the
+     * \ref RecordThreadState.add_variable function will create a new variable
+     * and mapping from the pointer to it.
+     */
+    void notify_free(const void *ptr) override;
+
+    ~RecordThreadState() {}
+
+    /**
+     * Adds an input of the recording.
+     * This is adds the slot of that variable to the \ref Recording.inputs
+     * vector.
+     */
+    void add_input(uint32_t input);
+    /**
+     * Adds an output to the recording.
+     * The output can be seen as a final operation, which also has to infer the
+     * size of its input variables.
+     * Therefore, we record the full \ref ParamInfo for each output variable.
+     */
+    void add_output(uint32_t output);
+
+    bool pause();
+    bool resume();
+
+    /// A helper scope, pausing recording.
+    struct pause_scope {
+        RecordThreadState *rts;
+        bool tmp;
+
+        pause_scope(RecordThreadState *rts) : rts(rts), tmp(rts->pause()) {}
+        ~pause_scope() { rts->m_paused = tmp; }
+    };
+
+    /// Is recording paused or has an exception been thrown?
+    /// Recording any operation should be gated by this function.
+    inline bool paused() { return m_paused || m_exception; }
+
+    /// Records an exception, thrown while recording an operation.
+    /// This is necessary to gracefully fail finishing freezing the function.
+    inline void record_exception() {
+        if (!m_exception)
+            m_exception = std::current_exception();
+    }
+
+    /**
+     * Record the Expand operation, corresponding to the `jitc_var_expand` call,
+     * with which the variable `index` has been expanded.
+     *
+     * Reductions in LLVM might be split into three operations.
+     * First the variable is expanded by its size times the number of workers +
+     * 1 Then the kernel writes into the expanded variable with some offset, and
+     * finally the variable is reduced.
+     * The expand operation allocates a new memory region and copies the old
+     * content into it.
+     * We catch this case if the input variable of a kernel has a reduce_op
+     * associated with it.
+     * The source variable is discovered by walking the operations to the last
+     * memcpy and memset, which are then disabled by this function.
```

Comment:
I think this explains it now:
```cpp
    /**
     * Record the Expand operation, corresponding to the ``jitc_var_expand``
     * call, with which the variable has been expanded.
     *
     * Reductions in LLVM might be split into three operations. First the
     * variable is expanded by its size times the number of workers + 1 Then the
     * kernel writes into the expanded variable with some offset, and finally
     * the variable is reduced. The expand operation allocates a new memory
     * region and copies the old content into it. We catch this case if the
     * input variable of a kernel has a ``reduce_op`` associated with it. The
     * source variable is discovered by walking the operations to the last
     * memcpy and memset, which are then disabled by this function. This has to
     * be done since these operations are replaced by the expand operation.
     */
```